File: src/cinegeist/convo/probes.py
"""Choose the next question by how much it teaches us, and decide when to stop asking.

This is the interesting part of the conversation, and it is maths, not prompting (plan.md §5.2).
Given the current profile, we score the genome-covered catalog, keep the high-scoring films still
in contention, and ask: which taste axis most *divides* that contested set? An axis where the
contenders all look the same teaches nothing; an axis that spreads them wide halves the search.
We pick the axis of maximum spread, weighted down where the profile is already confident, then
ground the question in two real films at opposite poles of that axis — the low-pole film chosen
to be otherwise as similar as possible to the high-pole one, so the contrast is about that axis
and little else. The LLM's only job, later, is to phrase the pair; the choice is made here.

The same code serves the cold open: with no profile yet, every film is in contention and the
axis of greatest spread across the catalog is the most discriminating first question.

Stopping is deterministic too (plan.md §5.3): stop when the top five stop moving, when we're
clearly confident, at a hard turn cap, or the moment the user asks — and an escape hatch is
offered on every turn regardless.
"""

from __future__ import annotations

import re
import sqlite3
from dataclasses import dataclass

import numpy as np

from ..catalog import genome

# How many of the highest-scoring films form the "contested" set an axis is judged against.
_POOL_TOP = 200
# An axis must spread the contested films at least this much (variance of relevance) to be worth
# asking about; below it, everything looks the same and there is no question to ask.
_MIN_SPREAD = 1e-4

# -- stopping-rule constants (provisional; they move to spec/ in session 5) ----------
MAX_TURNS = 9
MIN_TURNS = 2  # don't declare victory after a single answer (the user's stop always overrides)
STABLE_TURNS = 2  # stop once the top-5 is identical across this many most-recent turns
MARGIN_THRESHOLD = 0.15  # rank-1 minus rank-10 score gap that means "confident enough"

# Phrases that mean "stop quizzing me and show me something" — the escape hatch, honoured always.
_STOP_PATTERNS = (
    "just show me",
    "show me something",
    "just tell me",
    "just pick",
    "just recommend",
    "stop asking",
    "enough questions",
    "get on with it",
)


@dataclass(frozen=True)
class ProbeFilm:
    """One pole of a probe: a real film the user is asked to react to."""

    movie_id: int
    title: str
    year: int | None


@dataclass(frozen=True)
class Probe:
    """The next question: an axis to split, grounded in two films at its poles.

    ``film_high`` sits high on ``axis_name``; ``film_low`` sits low but is otherwise as close to
    ``film_high`` as the contested pool allows. ``question`` is a plain deterministic phrasing used
    in offline mode; online, the LLM rephrases it. Which film the user picks tells us the sign of
    their feeling on this axis.
    """

    axis_position: int
    axis_name: str
    spread: float
    film_high: ProbeFilm
    film_low: ProbeFilm
    question: str


@dataclass
class _Pool:
    ids: list[int]
    titles: list[str]
    years: list[int | None]
    vectors: np.ndarray  # (n_films, n_tags), the films' genome rows


def _load_pool(
    conn: sqlite3.Connection, matrix: np.ndarray, excluded: frozenset[int]
) -> _Pool | None:
    rows = conn.execute(
        "SELECT movie_id, clean_title, title, year, genome_row "
        "FROM movies WHERE genome_row IS NOT NULL ORDER BY genome_row"
    ).fetchall()
    ids: list[int] = []
    titles: list[str] = []
    years: list[int | None] = []
    genome_rows: list[int] = []
    for row in rows:
        if row["movie_id"] in excluded:
            continue
        ids.append(row["movie_id"])
        titles.append(row["clean_title"] or row["title"])
        years.append(row["year"])
        genome_rows.append(row["genome_row"])
    if not ids:
        return None
    vectors = np.asarray(matrix[genome_rows], dtype=np.float32)
    return _Pool(ids=ids, titles=titles, years=years, vectors=vectors)


def _default_uncertainty(profile_vector: np.ndarray, n_tags: int) -> np.ndarray:
    """Per-axis weight: axes the profile already has a strong opinion on are worth asking less.

    A proxy — the magnitude of the centroid on an axis stands in for how settled it is — so a
    caller with a real per-axis confidence can pass its own ``uncertainty`` instead.
    """
    if profile_vector is None or not profile_vector.any():
        return np.ones(n_tags, dtype=np.float32)
    return 1.0 - np.minimum(1.0, np.abs(profile_vector.astype(np.float32)))


def _tag_names(conn: sqlite3.Connection) -> dict[int, str]:
    return {
        row["position"]: row["name"]
        for row in conn.execute("SELECT position, name FROM genome_tags")
    }


def _ground_pair(contested: np.ndarray, position: int) -> tuple[int, int] | None:
    """Pick (high-pole index, low-pole index) within the contested set for one axis.

    The high pole is the film most strongly on the axis; the low pole is the film weakly on it
    that is otherwise closest to the high pole, so the two differ mainly on this axis.
    """
    relevance = contested[:, position]
    high = int(np.argmax(relevance))
    below = np.flatnonzero(relevance < relevance.mean())
    if below.size == 0:
        low = int(np.argmin(relevance))
    else:
        sims = genome.cosine_scores(contested[below], contested[high])
        low = int(below[int(np.argmax(sims))])
    if high == low:
        return None
    return high, low


def choose_probe(
    conn: sqlite3.Connection,
    matrix: np.ndarray,
    profile_vector: np.ndarray,
    *,
    excluded_movie_ids: frozenset[int] = frozenset(),
    asked_positions: frozenset[int] = frozenset(),
    uncertainty: np.ndarray | None = None,
    pool_top: int = _POOL_TOP,
) -> Probe | None:
    """Choose the most informative next probe, or ``None`` when nothing left divides the pool.

    With a zero ``profile_vector`` this is the cold open: the whole pool is contested and the axis
    of greatest spread across the catalog wins. Otherwise the contested set is the top ``pool_top``
    films by cosine to the profile, and the axis is chosen to spread *those*.
    """
    pool = _load_pool(conn, matrix, excluded_movie_ids)
    if pool is None:
        return None
    n_tags = int(matrix.shape[1])

    cold_start = not profile_vector.any()
    if cold_start or len(pool.ids) <= pool_top:
        contested_idx = np.arange(len(pool.ids))
    else:
        scores = genome.cosine_scores(pool.vectors, profile_vector)
        contested_idx = np.argsort(scores)[::-1][:pool_top]
    contested = pool.vectors[contested_idx]

    weight = (
        uncertainty if uncertainty is not None else _default_uncertainty(profile_vector, n_tags)
    )
    spread = np.var(contested, axis=0) * weight
    if asked_positions:
        spread[list(asked_positions)] = -1.0

    position = int(np.argmax(spread))
    if float(spread[position]) < _MIN_SPREAD:
        return None  # nothing divides the contenders any more

    grounded = _ground_pair(contested, position)
    if grounded is None:
        return None
    high_local, low_local = grounded
    high_idx = int(contested_idx[high_local])
    low_idx = int(contested_idx[low_local])

    names = _tag_names(conn)
    film_high = ProbeFilm(pool.ids[high_idx], pool.titles[high_idx], pool.years[high_idx])
    film_low = ProbeFilm(pool.ids[low_idx], pool.titles[low_idx], pool.years[low_idx])
    return Probe(
        axis_position=position,
        axis_name=names.get(position, f"tag#{position}"),
        spread=float(spread[position]),
        film_high=film_high,
        film_low=film_low,
        question=f"Which would you rather put on tonight — {film_high.title} or {film_low.title}?",
    )


# -- stopping rules ------------------------------------------------------------------


@dataclass(frozen=True)
class StopDecision:
    """Whether to stop asking, and why. ``reason`` is 'continue' when we keep going."""

    stop: bool
    reason: str


def wants_to_stop(text: str) -> bool:
    """True when the user's words are asking for the escape hatch ("just show me something")."""
    lowered = re.sub(r"\s+", " ", text.strip().lower())
    return any(pattern in lowered for pattern in _STOP_PATTERNS)


def should_stop(
    *,
    turn: int,
    top5_history: list[list[int]],
    top_scores: list[float] | None = None,
    user_requested: bool = False,
    max_turns: int = MAX_TURNS,
    min_turns: int = MIN_TURNS,
    stable_turns: int = STABLE_TURNS,
    margin_threshold: float = MARGIN_THRESHOLD,
) -> StopDecision:
    """Decide whether to stop, at the first rule that fires (plan.md §5.3).

    The user's request wins immediately. Otherwise we honour a hard turn cap, then — once past a
    small minimum so we don't stop after one answer — stop when the top five has held steady across
    the last ``stable_turns`` turns, or when rank 1 leads rank 10 by more than ``margin_threshold``.
    """
    if user_requested:
        return StopDecision(True, "user_request")
    if turn >= max_turns:
        return StopDecision(True, "max_turns")
    if turn < min_turns:
        return StopDecision(False, "continue")

    recent = top5_history[-stable_turns:]
    if len(recent) == stable_turns and all(snapshot == recent[0] for snapshot in recent):
        return StopDecision(True, "top5_stable")

    if top_scores is not None and len(top_scores) >= 10:
        if top_scores[0] - top_scores[9] >= margin_threshold:
            return StopDecision(True, "margin")

    return StopDecision(False, "continue")

File: web/src/probes.ts
// Choose the next question by how much it teaches us, and decide when to stop — a TypeScript port
// of `cinegeist.convo.probes` (plan.md §5.2, §5.3, §8.6). Two layers live here:
//
//   1. The spec-faithful algorithm — `chooseProbe`, `groundPair`, `shouldStop`, `wantsToStop` —
//      checked against scoring/probes.json, the same fixtures the Python suite asserts. These work
//      on full genome vectors, exactly like the CLI.
//
//   2. The demo's runtime selector — `selectDemoProbe` — which picks among the precomputed pairs in
//      probes.json by information gain over the contested shard pool. The shard ships SVD-compressed
//      taste vectors that can't be indexed by genome position, so the demo measures spread through
//      each film's top-tag table instead. It never sees coverage data (hard rule 9): selection is
//      by information gain alone, so the demo can't steer visitors into the shard's dense regions.

import { PROBES } from "./constants.ts";
import { cosineScores, rowSlice } from "./math.ts";
import { filmTopTags, type DecodedShard, type PrecomputedProbe } from "./shard.ts";

// -- the spec-faithful algorithm (mirrors convo/probes.py) ---------------------------

export interface ProbeMovie {
  movieId: number;
  genomeRow: number;
  title: string;
  year: number | null;
  vector: ArrayLike<number>;
}

export interface Probe {
  axisPosition: number;
  axisName: string;
  spread: number;
  filmHighId: number;
  filmLowId: number;
  question: string;
}

/** The deterministic phrasing of a this-or-that probe (PROBE_QUESTION_TEMPLATE, filled in). */
export function phrasePair(highTitle: string, lowTitle: string): string {
  return `Which would you rather put on tonight — ${highTitle} or ${lowTitle}?`;
}

function anyNonZero(v: ArrayLike<number>): boolean {
  for (let i = 0; i < v.length; i++) if (v[i] !== 0) return true;
  return false;
}

function argmax(v: ArrayLike<number>): number {
  let best = 0;
  for (let i = 1; i < v.length; i++) if (v[i] > v[best]) best = i;
  return best;
}

function argmin(v: ArrayLike<number>): number {
  let best = 0;
  for (let i = 1; i < v.length; i++) if (v[i] < v[best]) best = i;
  return best;
}

/** Indices of `scores` by descending value, ties by descending index (np.argsort[::-1]). */
function argsortDescending(scores: Float64Array): number[] {
  const idx = Array.from(scores, (_, i) => i);
  idx.sort((a, b) => scores[a] - scores[b] || a - b); // ascending, stable
  idx.reverse();
  return idx;
}

/** Population variance of each column of a flat, row-major `nCols`-wide matrix (np.var, ddof=0). */
function columnVariance(matrix: ArrayLike<number>, nCols: number): Float64Array {
  const rows = matrix.length / nCols;
  const mean = new Float64Array(nCols);
  for (let i = 0; i < rows; i++) for (let j = 0; j < nCols; j++) mean[j] += matrix[i * nCols + j];
  for (let j = 0; j < nCols; j++) mean[j] /= rows;
  const varr = new Float64Array(nCols);
  for (let i = 0; i < rows; i++)
    for (let j = 0; j < nCols; j++) {
      const d = matrix[i * nCols + j] - mean[j];
      varr[j] += d * d;
    }
  for (let j = 0; j < nCols; j++) varr[j] /= rows;
  return varr;
}

function defaultUncertainty(profile: ArrayLike<number>, nTags: number): Float64Array {
  const u = new Float64Array(nTags);
  if (!anyNonZero(profile)) {
    u.fill(1);
    return u;
  }
  for (let j = 0; j < nTags; j++) u[j] = 1 - Math.min(1, Math.abs(profile[j]));
  return u;
}

/** Pick (high-pole, low-pole) indices within a contested set for one axis; null if degenerate. */
export function groundPair(contested: ArrayLike<number>, nCols: number, position: number): [number, number] | null {
  const rows = contested.length / nCols;
  const relevance = new Float64Array(rows);
  for (let i = 0; i < rows; i++) relevance[i] = contested[i * nCols + position];
  const high = argmax(relevance);
  const mean = relevance.reduce((a, b) => a + b, 0) / rows;

  const below: number[] = [];
  for (let i = 0; i < rows; i++) if (relevance[i] < mean) below.push(i);

  let low: number;
  if (below.length === 0) {
    low = argmin(relevance);
  } else {
    const sub = new Float64Array(below.length * nCols);
    below.forEach((r, k) => {
      for (let j = 0; j < nCols; j++) sub[k * nCols + j] = contested[r * nCols + j];
    });
    const sims = cosineScores(sub, nCols, rowSlice(contested, nCols, high));
    low = below[argmax(sims)];
  }
  return high === low ? null : [high, low];
}

export interface ChooseProbeOptions {
  excluded?: Set<number>;
  askedPositions?: Set<number>;
  uncertainty?: ArrayLike<number> | null;
  poolTop?: number;
}

/** Choose the most informative next probe over full genome vectors, or null (plan.md §5.2). */
export function chooseProbe(
  movies: ProbeMovie[],
  nCols: number,
  profile: ArrayLike<number>,
  names: Map<number, string>,
  opts: ChooseProbeOptions = {},
): Probe | null {
  const excluded = opts.excluded ?? new Set<number>();
  const poolTop = opts.poolTop ?? PROBES.POOL_TOP;

  const pool = movies.filter((m) => !excluded.has(m.movieId)).sort((a, b) => a.genomeRow - b.genomeRow);
  if (pool.length === 0) return null;

  const poolVecs = new Float64Array(pool.length * nCols);
  pool.forEach((m, i) => {
    for (let j = 0; j < nCols; j++) poolVecs[i * nCols + j] = m.vector[j];
  });

  const coldStart = !anyNonZero(profile);
  let contestedIdx: number[];
  if (coldStart || pool.length <= poolTop) {
    contestedIdx = pool.map((_, i) => i);
  } else {
    const scores = cosineScores(poolVecs, nCols, profile);
    contestedIdx = argsortDescending(scores).slice(0, poolTop);
  }

  const contested = new Float64Array(contestedIdx.length * nCols);
  contestedIdx.forEach((p, i) => {
    for (let j = 0; j < nCols; j++) contested[i * nCols + j] = poolVecs[p * nCols + j];
  });

  const weight = opts.uncertainty ?? defaultUncertainty(profile, nCols);
  const spread = columnVariance(contested, nCols);
  for (let j = 0; j < nCols; j++) spread[j] *= weight[j];
  if (opts.askedPositions) for (const p of opts.askedPositions) spread[p] = -1;

  const position = argmax(spread);
  if (spread[position] < PROBES.MIN_SPREAD) return null;

  const grounded = groundPair(contested, nCols, position);
  if (grounded === null) return null;
  const [highLocal, lowLocal] = grounded;
  const filmHigh = pool[contestedIdx[highLocal]];
  const filmLow = pool[contestedIdx[lowLocal]];

  return {
    axisPosition: position,
    axisName: names.get(position) ?? `tag#${position}`,
    spread: spread[position],
    filmHighId: filmHigh.movieId,
    filmLowId: filmLow.movieId,
    question: phrasePair(filmHigh.title, filmLow.title),
  };
}

// -- stopping rules (mirror convo/probes.py) -----------------------------------------

export interface StopDecision {
  stop: boolean;
  reason: string;
}

// Phrases that mean "stop quizzing me and show me something" — the escape hatch, honoured always.
const STOP_PATTERNS = [
  "just show me",
  "show me something",
  "just tell me",
  "just pick",
  "just recommend",
  "stop asking",
  "enough questions",
  "get on with it",
];

/** True when the user's words ask for the escape hatch ("just show me something"). */
export function wantsToStop(text: string): boolean {
  const lowered = text.trim().toLowerCase().replace(/\s+/g, " ");
  return STOP_PATTERNS.some((p) => lowered.includes(p));
}

/** Decide whether to stop asking, at the first rule that fires (plan.md §5.3). */
export function shouldStop(args: {
  turn: number;
  top5History: number[][];
  topScores?: number[] | null;
  userRequested?: boolean;
  maxTurns?: number;
  minTurns?: number;
  stableTurns?: number;
  marginThreshold?: number;
}): StopDecision {
  const maxTurns = args.maxTurns ?? PROBES.MAX_TURNS;
  const minTurns = args.minTurns ?? PROBES.MIN_TURNS;
  const stableTurns = args.stableTurns ?? PROBES.STABLE_TURNS;
  const marginThreshold = args.marginThreshold ?? PROBES.MARGIN_THRESHOLD;

  if (args.userRequested) return { stop: true, reason: "user_request" };
  if (args.turn >= maxTurns) return { stop: true, reason: "max_turns" };
  if (args.turn < minTurns) return { stop: false, reason: "continue" };

  const recent = args.top5History.slice(-stableTurns);
  if (recent.length === stableTurns && recent.every((s) => arraysEqual(s, recent[0]))) {
    return { stop: true, reason: "top5_stable" };
  }
  const scores = args.topScores;
  if (scores && scores.length >= 10 && scores[0] - scores[9] >= marginThreshold) {
    return { stop: true, reason: "margin" };
  }
  return { stop: false, reason: "continue" };
}

function arraysEqual(a: number[], b: number[]): boolean {
  return a.length === b.length && a.every((x, i) => x === b[i]);
}

// -- the demo's runtime selector over precomputed probes (plan.md §8.2) --------------

/**
 * Choose the next precomputed probe to ask by information gain over the contested shard pool.
 *
 * Cold start (no reactions yet) scores every film, so the axis of greatest spread across the whole
 * shard wins — which is the highest-spread precomputed probe. As reactions sharpen the profile the
 * contested set narrows to the high-scoring films and the spread is recomputed over *those*, so the
 * question that best divides the current contenders is asked next. Relevance is read from each
 * film's top-tag table (the SVD vectors can't be indexed by genome position). Coverage is never
 * consulted. Returns null when nothing left divides the pool.
 */
export function selectDemoProbe(
  profile: Float64Array,
  shard: DecodedShard,
  probes: PrecomputedProbe[],
  opts: { askedAxes?: Set<number>; seenFilmIds?: Set<number>; poolTop?: number } = {},
): PrecomputedProbe | null {
  const askedAxes = opts.askedAxes ?? new Set<number>();
  const seen = opts.seenFilmIds ?? new Set<number>();
  const poolTop = opts.poolTop ?? PROBES.POOL_TOP;

  // The contested set: every film at cold start, else the top-scoring films by taste cosine.
  let contested: number[];
  if (!anyNonZero(profile)) {
    contested = shard.films.map((_, i) => i);
  } else {
    const cos = cosineScores(shard.vectors, shard.nComponents, profile);
    contested = argsortDescending(cos).slice(0, poolTop);
  }

  // Per contested film, its genome-tag relevance map (from the top-tag table).
  const relevanceMaps = contested.map((i) => {
    const m = new Map<number, number>();
    for (const t of filmTopTags(shard, i)) m.set(t.position, t.relevance);
    return m;
  });

  let best: PrecomputedProbe | null = null;
  let bestSpread = PROBES.MIN_SPREAD;
  for (const probe of probes) {
    if (askedAxes.has(probe.axis)) continue;
    if (seen.has(probe.high.id) || seen.has(probe.low.id)) continue;
    let mean = 0;
    for (const map of relevanceMaps) mean += map.get(probe.axis) ?? 0;
    mean /= relevanceMaps.length;
    let variance = 0;
    for (const map of relevanceMaps) {
      const d = (map.get(probe.axis) ?? 0) - mean;
      variance += d * d;
    }
    variance /= relevanceMaps.length;
    if (variance > bestSpread) {
      bestSpread = variance;
      best = probe;
    }
  }
  return best;
}
